Add temperature-based auto power options

timmmoore opened the pull request:
### Requirements
Current auto_e_fan only checks settarget temp of extruder, not sensor temp of extruder. So turns power off 30sec after printing while extruder is still hot.
auto_e_fan should check both settarget and actual sensor temp of extruder before switching power off. Similar for chamber auto fan.

### Description
is_power_needed() in power.cpp checks autofan_speed but the code that updates it in temperature.cpp is #if out. The #if just checks whether the extruder fan has a pin, should check whether it is configured for auto power e fan as well. Similar the chamber temp.
autofan_speed variables are used to track whether the fan should be on/off based on temp sensor, is used to turn fan on/off if it has a pin to control it. Auto power code also checks the variables but the variable is not updated if the fan is not pin controlled.
is_power_needed() also checks for controllerfan_speed but only if controller fan has a pin and then duplicates the code that updates controllerfan_speed, to check if the steppers have been stopped. So enable checking of controllerfan_speed and enable code that updates that variable and then remove duplicate code.


<!--

We must be able to understand your proposed change from this description. If we can't understand what the code will do from this description, the Pull Request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code recently, so please walk us through the concepts.

-->

### Benefits

Correct behavior of turning power off so extruder fan is not turned off before extruder has cooled down/
Similar behavior for chamber fan
Remove duplicate code for controller fan

<!-- What does this fix or improve? -->

### Related Issues

<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->
Bug 14378

timmmoore committed:
Fix to power supply off is delayed by e fan/steppers/chamber fan temp requirements

timmmoore committed:
update default and example configuraton*.h

thinkyhead commented:
Maybe this is not obvious, but `AUTO_POWER_CONTROLLERFAN` is actually dependent on `USE_CONTROLLER_FAN`. It is simply grouped with all the other `AUTO_POWER_*` options so that they are all together in one place.

To repeat: `AUTO_POWER_CONTROLLERFAN` should do absolutely nothing unless `USE_CONTROLLER_FAN` is enabled.

The same goes for the other `AUTO_POWER_*_FAN` options. If a particular fan is not present, then its auto-power option is simply ignored.

thinkyhead committed:
Undo some config changes

thinkyhead committed:
Revert changes, disable unused auto power options

thinkyhead commented:
I've reverted the extraneous changes based on my previous comments. Please re-check the code to see whether additional patches are needed to get the desired behavior.

timmmoore commented:
Thanks for the comments. Looking at the changes, AUTO_POWER_X_FANs assumes that any auto controlled fans has pins to control them but need power so if the fans are not pin controlled the AUTO_POWER_X_PIN defines not apply?
The case I have is the fans are directly connected to the controlled power. So not pin controlled fans but they turn on as soon as the auto power is supplied. The problem is they turn off before the extruder is down to the right temp because all is_power_needed() is looking at is the target temp not the extruder temp. What I was trying to do was to have is_power_needed() use the autofan_speed code so the main power is delayed being turned off until the extruder is down to the right temp.

thinkyhead commented:
Correct. At this time there is no support for using the external power supply as the means of turning on components that are not otherwise enabled. The power is being enabled to the board so that the firmware can then modulate the power going to the connected devices. If a device (e.g., a Chamber Fan controlled by Marlin) doesn't exist, I don't think it's going to work to use the power supply itself as a stand-in for the device.

thinkyhead commented:
> they turn off before the extruder is down to the right temp because all is_power_needed() is looking at is the target temp not the extruder temp

Well then that is the logic you would want to look into changing. It makes sense to me that the hotend auto fan should be using the larger of the current and target temperatures. But then, probably only the current temperature is needed, because PID doesn't start to kick in until you get close to the target.

timmmoore commented:
I found for the extruder fan I got the behavior I expected by checking the `autofan_speed` variable as if it was a controlled extruder fan and enabling the code in `temperature.cpp` to update the variable based on the temp sensor. The behavior was just want I expected. Which was the variable was set to 0 when the temp sensor went below fan off temp and then the power code did its timeout.

So this behavior is a new feature rather than a bug of existing feature?

If I add a new `#define` for this behavior rather than trying to override the existing `#define`s which have a different behavior, would this be the correct thing to do? and change it so it's a FR rather than a bug.

thinkyhead commented:
If you assign the auto fan pin to the same pin as the `POWER_PIN`, does the build fail?

thinkyhead commented:
> they turn off before the extruder is down to the right temp because all `is_power_needed()` is looking at is the target temp not the extruder temp

When there are no auto fans, it will fall through to that.

So, sure, other auto power options would be possible to fill in the gaps. 
```cpp
#define AUTO_POWER_HOTEND_C  60 // (°C) Keep power on for cooling
#define AUTO_POWER_CHAMBER_C 30 // (°C) Keep power on for cooling
```

timmmoore commented:
Yes it does build. Dont know if it will work, I will give it a try.

timmmoore committed:
Allow E Auto fan pin to equal PS_ON_PIN

timmmoore commented:
Can you take a look at what I just did. If E0_AUTO_FAN_PIN is #defined to PS_ON_PIN then it works.
I modified conditionals_post.h and temperature.cpp not to try to PWM the fan pin if its == to PS_ON_PIN.
I modified configuration.h and configuration_adv.h to enable POWER_SUPPLY and AUTO_POWER_CONTROL and E0_AUTO_FAN_PIN so you can see the config that would be allowed but they shouldn't be changed in the PR.

thinkyhead commented on Marlin/src/inc/Conditionals_post.h:
```diff
@@ -989,7 +989,14 @@
+#define HAS_AUTO_PIN_FAN_3 HAS_AUTO_FAN_3 && (E3_AUTO_FAN_PIN != PS_ON_PIN)
+#define HAS_AUTO_PIN_FAN_4 HAS_AUTO_FAN_4 && (E4_AUTO_FAN_PIN != PS_ON_PIN)
+#define HAS_AUTO_PIN_FAN_5 HAS_AUTO_FAN_5 && (E5_AUTO_FAN_PIN != PS_ON_PIN)
+#define HAS_AUTO_CHAMBER_PIN_FAN_0 HAS_AUTO_CHAMBER_FAN && (CHAMBER_AUTO_FAN_PIN != PS_ON_PIN)
```
It looks to me like you could just modify the definition of `HAS_AUTO_FAN_n`, etc., adding on `(En_AUTO_FAN_PIN != PS_ON_PIN)` and so on, and the behavior would be correct everywhere. No need to add these extra defines.

timmmoore replied:
Thanks, thats a good idea, I will test it out again later.

timmmoore replied:
Thanks, I did your suggestion and it works. Removed the changes to temperature.cpp for the extruder fans and chamber fans.
I did a similar change for the controller fan, I updated the macros in conditional_post.h with similar checks to the extruder fans. I did have to change to use the HAS_CONTROLER_FAN macro rather than USE_CONTROLER_FAN in a couple of files.
I compiled the code, with and without auto power and with and without autofan and with/without autofan pin = PS_ON_PIN.
Testing - I am using a delta - dont think it makes any difference.
1. Without this change, if I heat the extruder up ~80c, settarget to 0, wait 30seconds, extruder fan turned off and extruder temp was > 50C
2. With this change, heat extruder up ~80C, settarget to 0, wait 30sec, extruder fan is still running, 30sec after extruder temp < 50C power turns off and extruder fan stops.
3. With controllerfan, run axis home, power turns on, controllerfan on, 30 sec after homing finishes, power still on. disable steppers (M18), 30secs power turns off and controllerfan turns off.

timmmoore committed:
update from comments

timmmoore committed:
Similar change for controller fan to enable use of PS_ON_PIN

timmmoore committed:
Corrected auto chamber fan define

AnHardt commented:
> If you assign the auto fan pin to the same pin as the POWER_PIN, does the build fail?

Uahhhh!
Ich you want to cool down the extruder below autotemp, the complete power will be off after a while. Not a good idea.

timmmoore commented:
@AnHardt, not sure what you mean. On my system the PS_ON_PIN turns power on/off to heaters/fans/steppers. It doesn't turn power on/off to the CPU, usb, etc. So the AUTO_POWER turns heaters/etc. off until there is something to print, when I print, power turns on, and then turns power off again after everything cooled down. When I ask it to print again, everything powers up again.

AnHardt commented:
My comment was about this one single line, in the context of a normal not pwm autofan pin. All autofans and the power logic would work against each other. They have to be logicaly ored together.

By the way. Any autopower feature is inherently dangerous. The ps_on pin is our last rescue for the case of a shorted power MOSFET (what seems to be not that rare). This can only work if the default state is off and nothing but a human can switch it on.

timmmoore commented:
@Anhardt, thanks, I dont think what I am doing conflicts with the last rescue case, if you call minkill, power is turned off regardless of any temp sensors. Then the board is reset. So if power is completely removed, the board will shutdown before/after the reset depending on the power supply. In my case the board will reset and the board will be on but heaters/etc. off with a failed print so a human needs to intervene to fix the failed print and work out what happened.
There is a secondary behavior with the auto power. if you have auto power and a normal auto fan, the auto power is kept on until the auto fan is turned off. I wanted the same behavior when the auto fan is controlled by the auto power pin and this change does that but you are right that we can't have the auto power code and the auto fan code both trying to control the same pin.
The behavior I am looking at is if you set the auto fan pin to PS_ON_PIN than the auto fan code driving the auto fan pin is #if out and auto power controls the PS_ON_PIN. If the auto fan pin is anything else the current auto fan pin code is used. Regardless of the auto fan pin value the rest of the auto fan code is running which the auto power code uses to decide when to turn off.
This should mean the current auto fan behavior stays as is unless you set the auto fan pin to PS_ON_PIN. when auto power controls the pin, there isn't 2 pieces of code conflicting over the ownership of the PS_ON_PIN. Setting the auto pin to PS_ON_PIN is saying the fan is auto controlled but since the controlling pin is PS_ON_PIN the power system needs to control changing the pin and it uses the autofan variables to decide how to change it. That code was already there but only if the fan is auto controlled.
I have updated the PR to reflect this behavior but not tested it yet.

AnHardt commented:
We do support autostart files. Not using M80 there in is obvious. Any hidden automatism is a risk. For sure you will not remember all of them.

thinkyhead committed:
Update SanityCheck.h

thinkyhead commented:
To simplify (and to get rid of the need to toy with `PS_ON_PIN` needing to coincide with other pins as a "hint" to the firmware) I suggest we add these `AUTO_POWER_CONTROL` sub-options:
```cpp
// These use the "AUTO_FAN" temperature settings:
#define AUTO_POWER_E_COOLING  // Keep PSU on for E fans
#define AUTO_POWER_C_COOLING  // Keep PSU on for Chamber fans
```
Then `is_power_needed` can just have two extra bits of logic to check the current temperature readings, and that will be all.

thinkyhead committed:
Update more configs, revert default configs

thinkyhead committed:
initial re-implement (undo)

thinkyhead committed:
Tweak auto fan temperature handling

thinkyhead committed:
Apply new auto power control options

thinkyhead committed:
Add new options to example configs

thinkyhead commented:
Ok, this has been rebased and changed according to my previous comment. To get this into your working copy…
```
git fetch origin
git checkout Auto-Power-Fans
git reset --hard origin/Auto-Power-Fans
```
It seems a bit cleaner this way. See if it works for you.

timmmoore commented:
@thinkyhead , Thanks, looking through your changes. When I enable POWER_SUPPLY/AUTO_POWER_CONTROL/AUTO_POWER_E_TEMP the code in is_power_needed() to check autofan_speed variable is enabled but not the code in temperature.cpp/checkExtruderAutoFans() that sets the variable unless I set E0_AUTO_FAN_PIN, if I set it to PS_ON_PIN I end up with the problem of checkExtruderAutoFans() should not write to the pin.
I added to #define HAS_AUTO_FAN  (...  || ENABLED(AUTO_POWER_E_TEMP) || ENABLED(AUTO_POWER_C_TEMP))  that enables checkExtruderAutoFans() but doesn't enable writing to the E fan pins as long as I dont set the E or C fan pin.

timmmoore commented:
#thinkyhead, just noticed in the commited to bugfix-2.0x branch you handled the above case. Do ignore the above comments. I will test it later. Thanks

timmmoore commented:
thanks, looking at M105.cpp the code for checking for parameter 'R' hasn't made it into the bugfix-2.0.x branch from this branch. The other changes have made it.

thinkyhead commented:
> looking at `M105.cpp` the code for checking for parameter 'R' hasn't made it

That's the other PR, just merged.

timmmoore commented:
Thanks, just finished testing it and its working as expected.
